libs/ledger/src/chain/main_chain.cpp

```diff
@@ -537,31 +603,66 @@ void MainChain::WriteToFile()
       }
     }
 
-    // This block is now the head in our file
-    if (!failed)
+    if (failed)
     {
-      // store both the HEAD block as both its has and the "HEAD" entry
+      FETCH_LOG_WARN(LOGGING_NAME,
+                     "Failed to walk back the chain when writing to file! Block head: ",
+                     block_chain_.at(heaviest_.hash)->body.block_number);
+      return;
+    }
+
+    // This block will now become the head in our file
+    // Corner case - block is genesis
+    if (block->body.previous_hash == GENESIS_DIGEST)
+    {
+      FETCH_LOG_DEBUG(LOGGING_NAME, "Writing genesis. ");
+
       block_store_->Set(storage::ResourceAddress("head"), *block);
       block_store_->Set(storage::ResourceID(block->body.hash), *block);
+    }
+    else
+    {
+      FETCH_LOG_DEBUG(LOGGING_NAME, "Writing block. ", block->body.block_number);
 
-      FETCH_LOG_DEBUG(LOGGING_NAME, "Updating HEAD to ", ToBase64(block->body.hash));
+      // Recover the current head block from the file
+      IntBlockPtr current_file_head = std::make_shared<Block>();
+      IntBlockPtr block_head        = block;
+      block_store_->Get(storage::ResourceAddress("head"), *current_file_head);
 
-      // Walk down the file to check we have an unbroken chain
-      while (LookupBlockFromCache(block->body.previous_hash, block))
+      // Now keep adding the block and its prev to the file until we are certain the file contains
+      // an unbroken chain. Assuming that the current_file_head is unbroken we can write until we
+      // touch it or it's root.
+      for (;;)
       {
-        storage::ResourceID block_storage_key{block->body.hash};
+        block_store_->Set(storage::ResourceID(block->body.hash), *block);
 
-        // add the block to the file storage if it
-        if (!block_store_->Has(block_storage_key))
+        // Keep the current_file_head one block behind
+        while (current_file_head->body.block_number != block->body.block_number - 1)
         {
-          block_store_->Set(block_storage_key, *block);
+          block_store_->Get(storage::ResourceID(current_file_head->body.previous_hash),
+                            *current_file_head);
         }
 
-        // Clear the block from ram
-        FlushBlock(block);
+        // Successful case
+        if (current_file_head->body.hash == block->body.previous_hash)
```

Comment:
Perhaps it would be better to use rather block number to compare blocks -  it is more performant & more consistent with other code above.
```suggestion
        if (current_file_head->body.block_number == expected_block_num)
```